components/src/store/reducers/form.ts

```diff
@@ -0,0 +1,29 @@
+import { ActionTypes, TypeKeys } from "../actions/index";
+import { FormState } from "../../store/state";
+
+// Define the default values of form state.
+const getInitialState = (): FormState => {
+    return {
+        submittedFormIds: [],
+    };
+};
+
+// The preferences reducer. Note that it does not mutate state, but rather always returns a fresh copy.
+// https://redux.js.org/basics/reducers/
+export const form = (currentState = getInitialState(), action: ActionTypes): FormState => {
+
+    switch (action.type) {
+        case TypeKeys.ADD_SUBMITTED_FORM_ID:
+            const formId = action.key;
+
+            // Check to see if the user has submitted the form before. If not,
+            // add the formId.
+            if (currentState.submittedFormIds.indexOf(formId) === -1) {
+                currentState.submittedFormIds.push(action.key);
```

Comment:
So it's a subtle but important [Redux thing](https://redux.js.org/recipes/structuring-reducers/immutable-update-patterns), but the aim is to return a _copy_ of the state slice, rather than mutate it. So instead, what you'd want to do here is return a new object containing a clone of the current set of IDs and the newly submitted ID, de-duped, like

```
const submittedFormIds = currentState.submittedFormIds
    .concat(action.key)
    .reduce((ids, id) => ids.includes(id) ? ids : [...ids, id], []);

// Return a modified copy of currentState.
return Object.assign({}, { submittedFormIds });
```

(The `default` case just acts as a no-op.)